ext/net/net.rs

```diff
@@ -0,0 +1,79 @@
+use deno_core::anyhow::Context;
+use deno_core::error::AnyError;
+use fqdn::FQDN;
+use std::net::Ipv4Addr;
+use std::net::Ipv6Addr;
+use std::str::FromStr;
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub enum Host {
+  FQDN(FQDN),
+  Ipv4(Ipv4Addr),
+  Ipv6(Ipv6Addr),
+}
+
+impl Host {
+  pub fn from_str(host: &str, origin_host: &str) -> Result<Self, AnyError> {
+    if let Ok(ipv6) = host.parse::<Ipv6Addr>() {
+      return Ok(Host::Ipv6(ipv6));
+    }
+
+    let host = FQDN::from_str(host)
+      .with_context(|| format!("Failed to parse host: {}\n", origin_host))?;
+    let host_string = host.to_string();
+
+    if let Ok(ipv4) = host_string.parse::<Ipv4Addr>() {
+      return Ok(Host::Ipv4(ipv4));
+    }
+
+    return Ok(Host::FQDN(host));
+  }
+
+  pub fn to_string(&self) -> String {
+    match self {
+      Host::FQDN(fqdn) => fqdn.to_string(),
+      Host::Ipv4(ipv4) => ipv4.to_string(),
+      Host::Ipv6(ipv6) => format!("[{}]", ipv6.to_string()),
+    }
+  }
+}
+
+pub fn split_host_port(s: &str) -> Result<(String, Option<u16>), AnyError> {
+  let mut host = s.to_string();
+  let mut port = None;
+
+  let have_port = (&host).contains(":") && !(&host).contains("[");
+
+  if host.starts_with('[') && host.contains(']') {
+    if host.ends_with("]:") {
+      return Err(AnyError::msg("Invalid format: [ipv6]:port"));
+    }
+    if let Some(pos) = host.rfind("]:") {
+      let port_str = &host[pos + 2..];
+      let port_ = port_str.parse::<u16>().ok();
+      host = host[1..pos].to_string();
+      port = port_;
+    } else {
+      host = host[1..(host.len() - 1)].to_string();
+    }
+  } else if let Some(pos) = host.rfind(':') {
+    let port_str = &host[pos + 1..];
+    if let Ok(parsed_port) = port_str.parse::<u16>() {
+      host.truncate(pos);
+      port = Some(parsed_port);
+    }
+  }
+
+  if have_port && port.is_none() {
+    return Err(AnyError::msg("No port specified after ':'"));
+  }
+
+  Ok((host, port))
+}
+
+pub fn extract_host(s: &str) -> String {
```

Comment:
Is this meant only for URLs? If so maybe we should try to parse as `url::Url` and use appropriate methods from that crate? WDYT?

Reply:
i extract the host and the port from url to ask permission or give permission for this host at this port so we don't need to parse url itself
url it maybe  have  some  another  part like  path  of resource  

like 
https://github.com:443/denoland/deno

i  extract 'github.com:443'  so we need permission  for host:github.com at  port:443

so  i don't give any important  for resource of url ,


also we have  another  method  to  parse  url  it self  
'from_url'
but this  its needs valid url  valid  url 

so, i extracted the host from url to validate the  host  

like https://git@hub.com/denoland/deno
git@hub.com -> its invalid